scte35/segmentationdescriptor.go

```diff
@@ -0,0 +1,258 @@
+	segCloseRules[SegDescUnscheduledEventEnd][0x13] = segCloseNormal
+	segCloseRules[SegDescNetworkStart][0x13] = segCloseNormal
+	segCloseRules[SegDescNetworkEnd][0x13] = segCloseNormal
+}
+
+func (d *segmentationDescriptor) parseDescriptor(data []byte) error {
+	buf := bytes.NewBuffer(data)
+	// closure to ignore EOF error from buf.ReadByte().  We've already checked
+	// the length, we don't need to continually check it after every ReadByte
+	// call
+	readByte := func() byte {
+		b, _ := buf.ReadByte()
+		return b
+	}
+	if binary.BigEndian.Uint32(buf.Next(4)) != segDescID {
+		return mpegts.ErrSCTE35InvalidDescriptorID
+	}
+	d.eventID = binary.BigEndian.Uint32(buf.Next(4))
+	if readByte()&0x80 == 0 { // Cancel indicator
+		flags := readByte()
+		if flags&0x80 == 0 {
+			// skip over component info
+			ct := readByte()
+			if int(ct)*6 > buf.Len()-5 {
+				return mpegts.ErrInvalidSCTE35Length
+			}
+			for ; ct > 0; ct-- {
+				buf.Next(6)
+			}
+		}
+		d.hasDuration = flags&0x40 != 0
+		if d.hasDuration {
+			if buf.Len() < 10 {
+				return mpegts.ErrInvalidSCTE35Length
+			}
+			d.duration = uint40(buf.Next(5))
+		}
+		// upid unneeded now...
+		d.upidType = SegUPIDType(readByte())
+		upidLen := int(readByte())
+		if buf.Len() < upidLen+3 {
+			return mpegts.ErrInvalidSCTE35Length
+		}
+		d.upid = buf.Next(upidLen)
+		d.typeID = SegDescType(readByte())
+		d.segNum = readByte()
+		d.segsExpected = readByte()
+	}
+	return nil
+}
+
+func (d *segmentationDescriptor) SCTE35() SCTE35 {
+	return d.spliceInfo
+}
+
+func (d *segmentationDescriptor) EventID() uint32 {
+	return d.eventID
+}
+
+func (d *segmentationDescriptor) TypeID() SegDescType {
+	return d.typeID
+}
+
+func (d *segmentationDescriptor) IsOut() bool {
+	switch d.TypeID() {
+	case SegDescProgramStart, SegDescChapterStart,
+		SegDescProviderAdvertisementStart, SegDescDistributorAdvertisementStart,
+		SegDescProviderPOStart, SegDescDistributorPOStart,
+		SegDescUnscheduledEventStart, SegDescNetworkStart,
+		SegDescProgramOverlapStart, SegDescProgramResumption:
+		return true
+	default:
+		return false
+	}
+}
+
+func (d *segmentationDescriptor) IsIn() bool {
+	switch d.TypeID() {
+	case SegDescProgramEnd, SegDescChapterEnd,
+		SegDescProviderAdvertisementEnd, SegDescDistributorAdvertisementEnd,
+		SegDescProviderPOEnd, SegDescDistributorPOEnd,
+		SegDescUnscheduledEventEnd, SegDescNetworkEnd,
+		SegDescProgramRunoverPlanned, SegDescProgramRunoverUnplanned,
+		SegDescProgramEarlyTermination, SegDescProgramBreakaway:
+		return true
+	default:
+		return false
+	}
+}
+
+func (d *segmentationDescriptor) HasDuration() bool {
+	return d.hasDuration
+}
+
+func (d *segmentationDescriptor) Duration() mpegts.PTS {
+	return d.duration
+}
+
+func (d *segmentationDescriptor) UPIDType() SegUPIDType {
+	return d.upidType
+}
+
+func (d *segmentationDescriptor) UPID() []byte {
+	return d.upid
+}
+
+func (d *segmentationDescriptor) CanClose(out SegmentationDescriptor) bool {
```

Comment:
I don't want to go through and call out every instance of this, but in functions where we're passing around `SegmentationDescriptor` let's always do it by a reference.  There's no reason to take the memory hit.